### ISSUE TITLE
Compare tiling strategies for predict tile

### ISSUE THREAD

**Issue by bw4sz** [weecology/DeepForest#1053]
Let's start a discussion here for what would be optional for tiling strategies for main.predict_tile.

1. We want to cover the entire image.
2. We prefer fewer windows to more windows
3. We prefer smaller overlap among windows to reduce sensitity to NMS hyperparameters.
4. We want to reduce edge effects.
5. We prefer to not copy numpy arrays to avoid GPU memory usage.
6. We prefer threadsafe multiprocessing.

```from deepforest.utilities import Tiler, TiledImageimport torch
import numpy as np
from deepforest.preprocess import compute_windows

image = np.random.randint(0, 255, (400, 400, 3), dtype=np.uint8)
patch_size = 300
patch_overlap = 0.05

windows = compute_windows(image, patch_size, patch_overlap)

    def batch_tiler(self, input_tensor, size, overlap):
        """
        Creates overlapping views of a 4D tensor.

        Args:
            input_tensor (torch.Tensor): A 4D tensor of shape [N, C, H, W].
            size (int): The size of the sliding window (square).
            overlap (int): The overlap between adjacent windows.

        Returns:
            torch.Tensor: A tensor containing all the overlapping views.
                          The output shape is [N * num_windows, C, size, size].
        """

        # Get the input tensor shape
        N, C, H, W = input_tensor.shape

        # Calculate the number of padding pixels needed
        padding_h = size - (H % size)
        padding_w = size - (W % size)

        # Pad the input tensor
        padded_tensor = pad(input_tensor, (padding_w, padding_w, padding_h, padding_h), "constant", 0)

        # Calculate the step size for the sliding window
        step = size - overlap

        # Use torch.unfold to create the overlapping views
        unfolded_h = padded_tensor.unfold(2, size, step)
        unfolded = unfolded_h.unfold(3, size, step)

        # The unfolded tensor has shape:
        # [N, C, H', W', size, size]
        # where H' and W' are the number of sliding windows in the height and width dimensions

        # Reshape to [N * H' * W', C, size, size]
        output = unfolded.permute(0, 2, 3, 1, 4, 5).reshape(-1, C, size, size)
                
        return output

batch_windows = batch_tiler(image, patch_size, patch_overlap)

tiled_image = TiledImage(
                        image=image,
                        tile_size=patch_size,
                        overlap=patch_overlap*patch_size)

tiled_image.visualise(edges=True, boxes=True, midpoints=True)

# Print the windows
print("Batch Windows")
print(batch_windows)
print("--------------------------------")
print("Tiler Windows")
print([tile for tile in tiled_image.tiles])
print("--------------------------------")
print("Compute Windows")
print(windows)
 
```

**Comment by bw4sz** [comment]
This is the 'batch' strategy.

## Pros
- Threadsafe
-  doesn't rely on IO
- less sensitive to NMS strategies (which can be a big deal)
## Cons
- lots of padding and additional prediction for a tiny amount of image at the edges
- forces too many windows


![](https://files.slack.com/files-pri/TBYECD9MH-F08QY29FMFZ/image.png)

**Comment by jveitchmichaelis** [comment]
In the second case the Restor strategy would be to accept that you need 3 tiles in each direction but distribute them equally across the range so that the left and right edges align with the image. I'm not sure there's a strong argument either way, but one option would be to drop tiles that intersect less than some pixel threshold (say 10px).

**Comment by jveitchmichaelis** [comment]
This would be a similar default:

![Image](https://github.com/user-attachments/assets/61d68279-2d8c-4901-8efe-df2153bbbd91)

and you could shift by (-overhang/2, -overhang/2) to position the tiles centrally.
